ddprofiler/app/src/main/java/ddprofiler/preanalysis/PreAnalyzer.java

```diff
@@ -130,24 +146,66 @@ public Map<Attribute, Values> readRows(int num) {
 
     private void calculateDataTypes(Map<Attribute, List<String>> data) {
         // estimate data type for each attribute
-        for (Entry<Attribute, List<String>> e : data.entrySet()) {
-            Attribute a = e.getKey();
-            // Only if the type is not already known
-            if (!a.getColumnType().equals(AttributeType.UNKNOWN))
+        for (Entry<Attribute, List<String>> entry : data.entrySet()) {
+            Attribute attribute = entry.getKey();
+
+            // If the type is known, skip
+            if (!attribute.getColumnType().equals(AttributeType.UNKNOWN))
                 continue;
-            AttributeType aType;
+
+            AttributeType attributeType;
             if (pc.getBoolean(ProfilerConfig.EXPERIMENTAL)) {
                 // In experimental mode - force all data to be strings
-                aType = AttributeType.STRING;
+                attributeType = AttributeType.STRING;
             } else {
-                aType = typeOfValue(e.getValue());
+                attributeType = typeOfValue(entry.getValue());
             }
-            if (aType == null) {
+            if (attributeType == null) {
                 continue; // Means that data was dirty/anomaly, so skip value
             }
-            a.setColumnType(aType);
-            // a.setColumnType(AttributeType.STRING);
+            attribute.setColumnType(attributeType);
+        }
+    }
+
+    private void estimateSemanticTypes(Map<Attribute, List<String>> data) {
+        // estimate data type for each attribute
+        for (Entry<Attribute, List<String>> entry : data.entrySet()) {
+            Attribute attribute = entry.getKey();
+            if (attribute.getSemanticType() != null && !attribute.getSemanticType().isEmpty()) {
+                continue;
+            }
+            String semanticType = semanticTypeOfValue(entry.getValue());
+            attribute.setSemanticType(semanticType);
+        }
+    }
+
+    private String semanticTypeOfValue(List<String> values) {
+        Map<String, String> semanticTypes = new HashMap<>();
+        for (String value : values) {
+            String granularity = checkTemporalGranularity(value);
+            if (granularity != null) {
+                semanticTypes.put("type", "temporal");
+                semanticTypes.put("granularity", granularity);
+                try {
+                    return new ObjectMapper().writeValueAsString(semanticTypes);
```

Comment:
Based on my understanding, semanticType is designed as a dictionary including type and granularity. Why is there a need to serialize it to a String at this point? If it's for serialization purposes, I suggest moving the serialization logic to the place where the data profile is actually serialized.

Reply:
You are right, early serialization at this point is not needed. In fact, serialization of the dictionary can automatically done during profile store. Early serialization is removed at 44c3adbcb6fb3d07c972c2708c40dcd5207d91c9